src/storage-client/src/controller/rehydration.rs

```diff
@@ -378,11 +409,67 @@ where
                     Some(response) => response,
                 };
 
+                if let Err(e) = &response {
+                    self.set_internal_collection_statuses(InternalCollectionStatus::Unknown { error_details: Some(e.to_string()) }).await;
+                }
+
                 self.send_response(location, cluster_size, client, response)
             }
         }
     }
 
+    /// Sets the internal collection status for sources/sinks this task manages
+    async fn set_internal_collection_statuses(&self, status: InternalCollectionStatus) {
+        let (source_status_history_id, sink_status_history_id) = {
+            let introspection_ids = self.introspection_ids.lock().expect("poisoned lock");
+
+            (
+                introspection_ids[&IntrospectionType::SourceStatusHistory],
+                introspection_ids[&IntrospectionType::SinkStatusHistory],
+            )
+        };
+
+        self.collection_manager
+            .append_to_collection(
+                source_status_history_id,
+                self.sources
+                    .keys()
+                    .map(|id| {
+                        let row = crate::healthcheck::pack_status_row(
+                            *id,
+                            status.name(),
+                            status.error_details(),
+                            (self.now)(),
```

Comment:
seems like we might want a helper function somewhere that takes an `id`, a status, etc, and packs it and appends it. Its possible that helper should be a wrapper around the `collection_manager`. Also, it can hold the `Arc<Mutex>` with the introspection ids, so the rehydration module doesn't need to understand that

Reply:
Good callout, I added a wrapper struct to the `healthcheck` module. Let me know what you think! If it makes sense, I can go back and update references to `healthcheck::pack_status_row` in `controller` so we can make it private.